src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -248,9 +247,11 @@ private <T, C> void trainAndIndex(
         KNNVectorValues<T> knnVectorValues = vectorValuesRetriever.apply(vectorDataType, fieldInfo, VectorProcessingContext);
         QuantizationParams quantizationParams = quantizationService.getQuantizationParams(fieldInfo);
         QuantizationState quantizationState = null;
-        if (quantizationParams != null) {
+        // Count the docIds
+        int totalLiveDocs = getLiveDocs(vectorValuesRetriever.apply(vectorDataType, fieldInfo, VectorProcessingContext));
```

Comment:
Does this now mean there are 3 iterations for KNNVectorValues
1. For training
2. For counting docs
3. For indexing the values

Is this the only option we have?

Reply:
yes that is the only option we have.

Reply:
So time for doing this iteration is very cheap as we are not loading vectors. We do similar thing in BDV too: ref: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesReader.java#L76-L95

Also, for both 1,3 we read the vectors which is expensive operation but reading docIds is cheap. when I benchmarked sometime back it was not adding more than 1sec for 1M docs.